### PR TITLE
feat: vector store

### PR DESCRIPTION
**Pull request by TimPietrusky**
## Motivation

* Integrates a vector store (Qdrant)
* Runs Qdrant locally
* Starts the vector store when the app is started
* Added qdrant-win.7z into the installer

## Discuss

* Needs an OpenAI API Key for the embedding, which doesn't exist yet when this is started for the first time, so we need to come up with an idea on how to solve this once the API Key is there. Currently we just set the API Key "none"

### CODE REVIEW

File: .eslintrc.json
Comment (pixelass) [inline .eslintrc.json:1]: why is this needed?

File: jest.config.electron.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ const jestConfig = {
 	...defaults,
 	roots: ["<rootDir>/src/electron"],
 	testMatch: ["**/?(*.)test.ts"],
+	testPathIgnorePatterns: [".e2e."],
```
Comment (pixelass) [inline jest.config.electron.ts:7]: this suggests the naming pattern of e2e to be `*.e2e.test.ts`

File: jest.config.electron.e2e.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import { defaults } from "jest-config";
+
+const jestConfig = {
+	...defaults,
+	roots: ["<rootDir>/src/electron"],
+	testMatch: ["**/e2e/**/*.test.ts"],
```
Comment (pixelass) [inline jest.config.electron.e2e.ts:6]: shoul this rather be `*.e2e.test.ts`? (see comment beolw)

File: src/client/pages/[locale]/installer/02.tsx [hunk 3]
```diff
@@ -192,6 +192,10 @@ export default function Page(_properties: InferGetStaticPropsType<typeof getStat
 									url: "https://blibla-captain-assets.s3.eu-central-1.amazonaws.com/portable-git-win.7z",
 									destination: "portable-git",
 								},
+								{
```
Comment (pixelass) [inline src/client/pages/[locale]/installer/02.tsx:197]: How large is this file? Let's try to sort the files by size as this will slightly improve time used in unpacking.

Maybe we can add comments on expected file sizes like `~50MB` or `>4GB`.

File: src/electron/future/main.ts [hunk 4]
```diff
@@ -242,4 +244,14 @@ export async function main() {
 			await createMainWindow();
 		});
 	}
+
+	const apiKey = keyStore.get("openAiApiKey") || "none";
```
Comment (pixelass) [inline src/electron/future/main.ts]: you can add a second argument as default. Also I suggest using an emptyt string as it will pass boolean checks.

```suggestion
	const apiKey = keyStore.get("openAiApiKey", "");
```

File: src/electron/future/services/vector-store.ts [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+
+		return VectorStore.instance;
+	}
+
+	/**
+	 * Inserts or updates documents in the specified collection. If a document ID is not provided,
+	 * the method will attempt to find an existing document based on the payload 'id' and 'language'.
+	 * If an existing document is not found, a new UUID will be generated.
+	 * 
+	 * Example: Auto-ID (recommended)
```
Comment (pixelass) [inline src/electron/future/services/vector-store.ts]: example snippets should use the `@example` directive

File: src/electron/future/utils/path-helpers.ts [hunk 6]
```diff
@@ -12,6 +12,8 @@ export const resourcesDirectory = isDevelopment
 	? path.join(process.cwd(), "resources")
 	: path.join(app.getPath("exe"), "..", "resources", "app.asar.unpacked", "resources");
 
+export const userDataDirectory = isDevelopment ? process.cwd() : app.getPath("userData");
```
Comment (pixelass) [inline src/electron/future/utils/path-helpers.ts]: can we add a dedicated folder here which we can then ignore. This approach won't scale as it could poolute the root directory

Comment (pixelass) [reply src/electron/future/utils/path-helpers.ts]: actually, userData is already handled. it uses either `captain` or `captain__development__`

image

File: src/electron/future/utils/path-helpers.ts [hunk 7]
```diff
@@ -25,7 +27,7 @@ export function getDirectory(...subpath: string[]): string {
 }
 
 export function getUserData(...subpath: string[]): string {
-	return path.join(app.getPath("userData"), ...subpath);
+	return path.join(userDataDirectory, ...subpath);
```
Comment (pixelass) [inline src/electron/future/utils/path-helpers.ts]: see above. We cannot do this.

**Review by pixelass — CHANGES_REQUESTED** [review]
Please see comments

File: .eslintrc.json
Comment (TimPietrusky) [reply .eslintrc.json:1]: This was needed to have "e2e" as part of the test-name. But I decided against it and moved it into a folder, where eslint is not complaining about the folderName. 

Removed this change again.

File: jest.config.electron.ts (hunk 1 above)
Comment (TimPietrusky) [reply jest.config.electron.ts:7]: This is a regex pattern, which makes sure to work for both directories like `__test__/e2e/*` and files `*.e2e.test.ts`.

Comment (TimPietrusky) [reply jest.config.electron.ts:7]: And the purpose is to **ignore** any files / folders that contain `e2e`.

File: jest.config.electron.e2e.ts (hunk 2 above)
Comment (TimPietrusky) [reply jest.config.electron.e2e.ts:6]: I decided against having `e2e` in the file name, but have a dedicated folder for e2e-tests.

File: src/client/pages/[locale]/installer/02.tsx (hunk 3 above)
Comment (TimPietrusky) [reply src/client/pages/[locale]/installer/02.tsx:197]: The files are already sorted by size (huge > small). I added the `size`, so we also see this.

File: src/electron/future/services/vector-store.ts (hunk 5 above)
Comment (TimPietrusky) [reply src/electron/future/services/vector-store.ts]: Thanks, those have been fixed.

File: src/electron/future/utils/path-helpers.ts (hunk 6 above)
Comment (TimPietrusky) [reply src/electron/future/utils/path-helpers.ts]: My problem here is, that I can't use `app.getPath`, as this is something that only exists in electron. For testing I want to run my e2e tests without electron, as this allows me to directly execute everything and don't interact with my code via IPC. So both of these folders are not reachable for my tests at all. 

This approach is also not poluting the root, as the folder is .gitignored, so it will only be relevant for a person that wants to test this. 

This solution was something I came up with after trying many other things and all of them failed. But I will try one last thing and mock the app.getPath folder for this specific test.

Comment (TimPietrusky) [reply src/electron/future/utils/path-helpers.ts]: It worked, I mocked `app.getPath` with `process.cwd()` for my specific test. I removed all the other changes to `path-helpers`.

File: src/electron/future/utils/path-helpers.ts (hunk 7 above)
Comment (TimPietrusky) [reply src/electron/future/utils/path-helpers.ts]: Was removed, see explanation above.

File: src/electron/future/services/vector-store.ts [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+	 * ```javascript
+	 * const document = {
+	 *   id: v4(),
+	 *   content: "Live Painting is very nice",
+	 *   payload: {
+	 *     id: "live-painting:schema",
+	 *     language: "en",
+	 *   },
+	 * };
+	 * ```
```
Comment (pixelass) [inline src/electron/future/services/vector-store.ts:186]:
```suggestion
	 * @example Auto-ID (recommended)
	 * const document = {
	 *   content: "Live Painting is very nice",
	 *   payload: {
	 *     id: "live-painting:schema",
	 *     language: "en",
	 *   },
	 * };
	 *
	 * @example Custom-ID using uuid
	 * const document = {
	 *   id: v4(),
	 *   content: "Live Painting is very nice",
	 *   payload: {
	 *     id: "live-painting:schema",
	 *     language: "en",
	 *   },
	 * };
```

**Review by pixelass — CHANGES_REQUESTED** [review]

**Review by pixelass — APPROVED** [review]

File: src/electron/future/services/vector-store.ts (hunk 8 above)
Comment (TimPietrusky) [reply src/electron/future/services/vector-store.ts:186]: I will stick with using the current formatting, as this is the default https://tsdoc.org/pages/tags/example/
